### DISCUSSION TITLE
Are there any resources that take this approach with other languages/concepts?

### DISCUSSION THREAD

**Discussion by andrew-werdna** [quii/learn-go-with-tests#634] [Q&A]
I love this project, and am very proud to be a sponsor. I would also love to see this paradigm flourish and become more widespread. Does anyone know of any other `learn x with tests` style of git repositories or other forms of resources with other programming languages?

I would love to see a `learn rust with tests`.

This approach to learning a programming language seems novel and more effective/complete to me for several reasons that I struggle to articulate concisely. 

This doesn't just stop at the most trivial testing scenarios/examples. It seems (IMO) to not decrease the rigor/depth/importance applied to the testing/verifiability mentality and strategies throughout. I think it's a solid introduction to multiple, basic good practices and Go testing idioms, covers a robust set of common use-cases/scenarios, and it just keeps improving.

Sometimes doing TDD and creating valuable tests/assertions across the different kinds of tests can be tricky. This repository seems to put in one place, many testing idioms/patterns that took me a little while to learn with Go, and I had to do a lot of different Google searches to discover many of them.

**Comment by quii** [comment]
First off, glad you like it, and thanks for the sponsorship. 

I don't know why, but I never get notifications on discussions! 

I would love to learn Rust enough to write the book, but I suspect unless I find a fulltime job doing Rust, that doesn't seem likely. I'd be at risk of doing an expert beginner guide. 

It's not impossible I'll be taking on a Kotlin role soon, in which case maybe that could happen. I think for my own career development though, I'd quite likely approach a publisher to get it turned into a "real" book. Of course, this Go version I'll continue to support and add more content when I can!

**Reply by andrew-werdna** [reply]
Understood, and that completely makes sense. My company actually just had massive layoffs (including my entire division), so my former teammates and I are commiserating on Discord and helping each other source roles. I've been lucky to encounter two separate job postings that are like 80% Golang and 20% Rust. That to me sounds very appealing as I'd love to learn Rust but continue using Go as well.

**Reply by andrew-werdna** [reply]
If it's acceptable to you, I'd love to leave this question open, just in case others find this thread and share resources (particularly with Rust).

**Comment by quii** [comment]
Elixir is quite a fun and interesting language, someone is making an attempt here https://angelmz.gitbook.io/learn-elixir-with-tests

**Reply by andrew-werdna** [reply]
Ah! Thank you for this! I'm also interested in Elixir (and especially being able to leverage the BEAM VM).

**Comment by andrew-werdna** [comment]
@quii I gotta say, I'm an even bigger fan since I found this little gem:
 https://github.com/learn-with-tests/learn-rust-with-tests

It looks like you're contributing to that as well? 

I'm pleased to say I started a new job about 3-4 months ago in which I was told I'd be doing Golang and Rust. To my surprise, I've been doing about 95% Rust, and only now have we been adding some Golang work. I've been eating and sleeping Rust since the end of March of this year! I'm by no means an expert of course, but I've grown by leaps and bounds. I'd love to actually contribute to that project however I can. 

I saw the option to buy a coffee, but I was wondering if it would be more helpful to just increase my GitHub sponsorship of you by several $/month as well as try to contribute to that other repository. What are your thoughts on the $ side? 

You've been doing the good work for a while now, and I'm finally in a position to potentially help out in a more meaningful way!

**Reply by andrew-werdna** [reply]
Total aside, but I've become quite fascinated with formal methods, specification, and modeling. I've read quite a bit about, and played around with [Alloy 6](https://haslab.github.io/formal-software-design/) now that it has temporal logic support. It still seems like [TLA+](https://learntla.com/) is the industrial strength OG in many ways. I've also heard a bit about the [P framework](https://p-org.github.io/P/) as well.

Since you're (at least somewhat) into Rust now, are you familiar with [stateright](https://www.stateright.rs/)? It looks really cool and the book has a direct comparison with TLA+!

**Reply by quii** [reply]
Hey

Im glad you've found this all useful. Learn rust with tests was a thing I attempted to start with some friends a few years ago (look at the commit dates). Sadly, like how lots of side-projects go, it didn't. Whilst I would love to spend some time learning rust enough, and having the time to write an equivalent to this in Rust, at this point at least, I just don't have the time. 

Not to say it's out of the question, I'd love to do it one day, but I would need _significant_ sponsorship to be able to quit my job and do it! 

Thanks again for the kind words.

**Reply by andrew-werdna** [reply]
I was chatting with one of my co-workers about this today. We are not experts, but we have lived Rust all day for the last 4-6 months. Would you be open to potential contributions?
